### PR TITLE
Wallet support

### PR DESCRIPTION
**Pull request by dkent600**
Resolves https://daostack.tpondemand.com/RestUI/Board.aspx#page=board/5209716961861964288&appConfig=eyJhY2lkIjoiQjgzMTMzNDczNzlCMUI5QUE0RUE1NUVEOUQyQzdFNkIifQ==&boardPopup=feature/1156/silent

- supports extension wallets (like MetaMask and Safe) as well as Portis and Fortmatic
- Adds Login and Logout buttons in the user account dropdown
- The login button, rather than showing the entire list of providers, reloads cached provider information to automatically log the user into their previously-selected (and therefore cached) provider

**Notes**
- there are lots of permutations here even when not accounting for support for new providers.  Test thorougly!
- see the comment I made below about the login/logout icons

### CODE REVIEW

**Comment by dkent600** [comment]
@A-Zak   I think the logout icon you gave me is more appropriate as a login icon, like the metaphor given in these:

Login like this:  
image


And a logout icon would be like this:

image

**Comment by leviadam** [comment]
Maybe one small comment @dkent600.
The login text inside the hover has a MM logo next to it, even when you know it is not a MM account, which is very strange.
Say the storage knows that I'm a Portis user, and it shows my account, but I still have a log in button with a MM logo, it is very confusing.

Either not show a logo there, or show the correct one.

File: src/layouts/AppContainer.tsx [hunk 1]
```diff
@@ -64,59 +66,88 @@ class AppContainer extends React.Component<IProps, IState> {
...
+      } else {
+        console.log("failed to instantiate cached web3Provider");
+        this.uncacheWeb3Info();
+      }
     }
+    return false;
   }
 
-  public render() {
+  public render(): any {
```
Comment (tibetsprague) [inline src/layouts/AppContainer.tsx]: yeah im over this lint rule

**Comment by dkent600** [comment]
@leviadam   @A-Zak 

> Either not show a logo there, or show the correct one.

I totally agree, but it is my understanding that this is Alex's call.

File: src/arc.ts [hunk 2]
```diff
@@ -37,9 +73,58 @@ const settings = {
...
+function getWeb3(): any {
+  const arc = (window as any).arc;
+  const web3 = arc ? arc.web3 : null;
+  return web3;
+}
+
+/**
+ * Return the default account, locked, in current use by Arc.
+ */
+async function _getCurrentLockedAccount(): Promise<string> {
```
[the remark this answers is not in the document]
Comment (dkent600) [reply src/arc.ts]: The assumption is that `account[0]` is only available when locked.

File: src/arc.ts [hunk 3]
```diff
@@ -101,128 +187,287 @@ export async function checkWeb3Provider() {
...
-  // console.log(arcSettings);
-  // console.log(`alchemy-server (process.env.API_URL): ${process.env.API_URL}`);
-  if (arcSettings.web3Provider.isSafe) {
-    console.log("Using Gnosis Safe");
-  } else if (arcSettings.web3Provider.isMetaMask) {
-    console.log("Using MetaMask");
-  } else {
-    console.log("Using default Web3 (read-only) provider");
+/**
+ * @return the currently-locked account address from Arc. Ignores any injected
```
[the remark this answers is not in the document]
Comment (dkent600) [reply src/arc.ts]: Could also be I'm also not using the term correctly.  By "locked" I mean that there is an account that can be used to sign transactions.

File: src/layouts/AppContainer.tsx (hunk 1 above)
Comment (dkent600) [reply src/layouts/AppContainer.tsx]: Not me, at all.  'any' is really not a big deal.  Not having the the non-trivial return values is.

File: src/arc.ts [hunk 4]
```diff
@@ -37,9 +73,58 @@ const settings = {
...
+    case "staging" : {
+      arcSettings = settings.staging;
+      break;
+    }
+    case "production" : {
+      arcSettings = settings.production;
+      break;
+    }
+    default: {
+      console.log(process.env.NODE_ENV === "development");
```
[the remark this answers is not in the document]
Comment (dkent600) [reply src/arc.ts]: Done

File: src/arc.ts [hunk 5]
```diff
@@ -1,7 +1,43 @@
...
+      {
+        portis: {
+          id: "aae9cff5-6e61-4b68-82dc-31a5a46c4a86",
+          network: "mainnet",
+        },
+        fortmatic: {
+          key: "pk_live_38A2BD2B1D4E9912",
+        },
+      }
+      : // (process.env.NODE_ENV === "staging") ?
```
[the remark this answers is not in the document]
Comment (dkent600) [reply src/arc.ts]: Yeah, I guess it's OK to hide Portis and Fortmatic when in development mode.

**Comment by leviadam** [comment]
> @leviadam @A-Zak
> 
> > Either not show a logo there, or show the correct one.
> 
> I totally agree, but it is my understanding that this is Alex's call.

Just so that we are on the same page here. This is not the same logo of MM we discussed before. Before we talked about the one shown if there is no info, and when you click on it, you get the web3connect options. Now I'm talking about the one inside the hover, in which you already know what is the provider.

File: src/arc.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (tibetsprague) [reply src/arc.ts]: gotcha. where does that term come from?

[the remark this answers is not in the document]
Comment (leviadam) [reply src/arc.ts]: "Locked" does not sound like a good term for that. Active maybe?

File: src/layouts/AppContainer.tsx (hunk 1 above)
Comment (leviadam) [reply src/layouts/AppContainer.tsx]: I think if both @jellegerbrandy and @tibetsprague prefer not to have this rule we should turn it off @dkent600.

**Comment by dkent600** [comment]
@leviadam 

> > > Either not show a logo there, or show the correct one.
> > 
> > I totally agree, but it is my understanding that this is Alex's call.
> Just so that we are on the same page here. This is not the same logo of MM we discussed before. 

See my comments above about the icons.  That is exactly what I was referring to. :-)

File: src/arc.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (jellegerbrandy) [reply src/arc.ts]: but why would it be *locked"? Is that even relevent here? It is a  confsing name, why not rename it "getCurrentAccountFromProvider"?

File: src/arc.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (jellegerbrandy) [reply src/arc.ts]: we can just drop the adjective. "Current acocunt" instead of "currently-locked account"?

**Comment by jellegerbrandy** [comment]
I am having some problems with the review app: https://alchemy-staging-rinkeby-pr-854.herokuapp.com

Basically, the Metamask button on the "log in" screen is unresponsive, and I see this in the console:
* 
image

* 
image

**Comment by jellegerbrandy** [comment]
Clicking on the fortmatic link gives me this: 
image

**Comment by jellegerbrandy** [comment]
I also get lots of console errors trying to login to portis (with a non-existing account) that are not reported in the UX: 
image

**Comment by jellegerbrandy** [comment]
image

**Comment by dkent600** [comment]
@jellegerbrandy

The unique key errors exist in both Alchemy and Web3Connect.  I have already created tickets on both these errors in the respective repos.

**Comment by dkent600** [comment]
> the Metamask button on the "log in" screen is unresponsive

It works fine for me.  I could use a repro case on this (steps to reproduce).

**Comment by A-Zak** [comment]
@dkent600 let's use the icons you found for login and logout...
and for the main login button, let's just remove the MM icon...

**Comment by dkent600** [comment]
@jellegerbrandy 

> Clicking on the fortmatic link gives me this:

Fortmatic works fine for me.  Do you have a fortmatic account?

> I also get lots of console errors trying to login to portis (with a non-existing account)

Same for Portis, no problem for me.

I think I need better repro cases here, like the steps that occur when there is no account.

**Comment by dkent600** [comment]
@jellegerbrandy 

> the Metamask button on the "log in" screen is unresponsive

I am guessing you are connected to the wrong network, and we aren't showing the error message except in the console. I will fix this.

**Comment by dkent600** [comment]
@jellegerbrandy I have done some refactoring that cleans up some code and improves the logging and error handling/reporting.  If your problem with MM was due to being connected to the wrong network, you should now see a notification to that effect.

**Comment by dkent600** [comment]
@jellegerbrandy @A-Zak @leviadam @tibetsprague   With today's dev meeting conversation in mind (about letting the user know a little more about what is happening), I added notifications for when logging in and lgging out.

Also updated the icons.

**Comment by dkent600** [comment]
@jellegerbrandy  I am not having any problem with Portis nor Fortmatic with no account (though honestly I'm not sure what you mean by that, I just used incognito mode  -- need a fleshed-out repro case)

**Pull request merged**